### PR TITLE
Initial implementation of ReadOnly structs

### PR DESCRIPTION
**Pull request by VSadov**
Supporting parsing, binding, emit.

UNDONE: we are not emitting "IsReadOnly" attribute yes, so metadata scenarios do not work yet.

The feature is very simple.  We now support the following syntax:

```C#
readonly struct S1
{
}
```

It means that "this" parameters of all instance methods, with exception of constructors, are effectively `ref readonly`

The most observable result is that invoking methods of such struct on a readonly instance does not require copying that said instance.

### CODE REVIEW

**Comment by VSadov** [comment]
@dotnet-bot test windows_build_correctness_prtest  please

**Comment by VSadov** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler  - please take a look

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 1]
```diff
@@ -398,7 +398,13 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
 
                     // SPEC: When this is used in a primary-expression within an instance method or instance accessor
                     // SPEC: of a struct, it is classified as a variable. 
-                    if (!thisref.Type.IsValueType || RequiresReturnableReference(valueKind))
+
+                    // "this" is only readable if we are not in a struct. RValueOnly is checked at the beginning of this method
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Should fix this comment. Now that I read it seems ambiguous. The comment should say: If we are not in a struct "this" is only readable. Since we are hear, we need more than readable - RValueOnly is checked at the beginning of this method.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 2]
```diff
@@ -398,7 +398,13 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
...
+                    if (!thisref.Type.IsValueType ||
+                        //"this" is only readable in members of readonly structs, uless weare in a constructor
+                        (thisref.Type.IsReadOnly && (this.ContainingMember() as MethodSymbol)?.MethodKind != MethodKind.Constructor) ||
+                        //"this" is not returnable by reference in a struct
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Consider one comment on the if to describe the scenario vs. commenting every line.  IMHO breaks up the code, makes it harder to follow.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit.Abstractions;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Parsing
+{
+    [CompilerTrait(CompilerFeature.ReadonlyReferences)]
+    public class RadOnlyStructs : ParsingTests
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs]: RadOnly?  Did you mean ReadOnly?  Same for the file name.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+using Xunit;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit.Abstractions;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Parsing
+{
+    [CompilerTrait(CompilerFeature.ReadonlyReferences)]
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs]: The OCD in me can't help but mention the casing inconsistency here:

- `CompilerFeature.ReadonlyReferences`
- `class ReadOnlyStructs`

Prefer consistent casing here around `ReadOnly`.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 2 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: I kind of thought that commenting on different conditions separately would be more readable. Perhaps not.

---
In reply to: [118026950](https://github.com/dotnet/roslyn/pull/19713#discussion_r118026950) [](ancestors = 118026950)

File: src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs (hunk 4 above)
Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs]: Will fix the trait to be "ReadOnlyReferences"

---
In reply to: [118027328](https://github.com/dotnet/roslyn/pull/19713#discussion_r118027328) [](ancestors = 118027328)

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - PING. I think I addressed al the concerns.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 5]
```diff
@@ -398,7 +398,13 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
 
                     // SPEC: When this is used in a primary-expression within an instance method or instance accessor
                     // SPEC: of a struct, it is classified as a variable. 
-                    if (!thisref.Type.IsValueType || RequiresReturnableReference(valueKind))
+
+                    // Note: RValueOnly is checked at the beginning of this method. Since we are here we need more that readable.
+                    //"this" is only readable in members of readonly structs, uless weare in a constructor.
+                    //"this" is not returnable by reference in a struct.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Typos: `more than readable`, `unless we are`?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+        {            
+            x = i; // OK
+        }
+
+        public S1(S1 arg)
+        {
+            this = arg; // OK
+        }
+
+        public void Tes1()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs]: Typo: `Test1`

File: src/Compilers/CSharp/Portable/CodeGen/EmitAddress.cs [hunk 7]
```diff
@@ -111,6 +111,19 @@ private LocalDefinition EmitAddress(BoundExpression expression, AddressKind addr
...
+                //PROTOTYPE(readonly-ref): the following is more efficient codegen
+                //                         not enabling in this change since results in too much of test churn and is not required for the current change
+                //                         should be enabled together with other similar "PROTOTYPE" in codegen.
+                //case BoundKind.DefaultOperator:
+                //    var type = expression.Type;
+
+                //    var temp = this.AllocateTemp(type, expression.Syntax);
+                //    _builder.EmitLocalAddress(temp);                  //  ldloca temp
+                //    _builder.EmitOpCode(ILOpCode.Dup);                //  dup
+                //    _builder.EmitOpCode(ILOpCode.Initobj);            //  intitobj  <type>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CodeGen/EmitAddress.cs]: Typo: `initobj`

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - PING

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 8]
```diff
@@ -1412,26 +1412,38 @@ private void EmitCallExpression(BoundCall call, UseKind useKind)
...
-                        Debug.Assert(receiverType == methodContainingType);
-                        // method is defined in the struct itself and is assumed to be mutating.
-                        //PROTOTYPE(readonlyRefs): when readonly structs are implemented, this will have to check "this"
-                        tempOpt = EmitReceiverRef(receiver, AddressKind.Writeable);
-                        callKind = CallKind.Call;
+                        // if method is defined in the struct itself it is assumed to be mutating, unless 
+                        // it is a member of a readonly struct and is not a constructor
+                        var thisAddresskind = methodContainingType.IsReadOnly && method.MethodKind != MethodKind.Constructor ?
+                                                                        AddressKind.ReadOnly :
+                                                                        AddressKind.Writeable;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: Doesn't `methodContainingType` refer to the type of the method being invoked vs. the type of the method we are presently inside of?

**Comment by jaredpar** [comment]
@VSadov looking

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs [hunk 9]
```diff
@@ -194,18 +196,20 @@ internal static LanguageVersion RequiredVersion(this MessageID feature)
             // Checks are in the LanguageParser unless otherwise noted.
             switch (feature)
             {
+                //PROTOTYPE(readonlyRefs): these are all 7.2 features we need to update this.
+                // C# 7.1 features.
+                case MessageID.IDS_FeatureReadonlyReferences:
+                case MessageID.IDS_FeatureRefStructs:
+                case MessageID.IDS_FeatureReadonlyStructs:
+                    return LanguageVersion.CSharp7_1;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs]: I thought @OmarTawfik already added CSharp7_2.

File: src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs [hunk 10]
```diff
@@ -40,10 +40,22 @@ public override RefKind RefKind
...
-                return
-                    ((object)_containingType == null || _containingType.TypeKind != TypeKind.Struct) ? RefKind.None :
-                    ((object)_containingMethod != null && _containingMethod.MethodKind == MethodKind.Constructor) ? RefKind.Out :
-                    RefKind.Ref;
+                if (ContainingType?.TypeKind != TypeKind.Struct)
+                {
+                    return RefKind.None;
+                }
+
+                if (_containingMethod?.MethodKind == MethodKind.Constructor)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs]: What will the value of `_containingMethod` be inside a local function which is inside a constructor?  Same for lambdas. 

Can you point me to the test which covers these cases?

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs (hunk 9 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs]: Evidently the features are still marked as 7_1. 
I did not want to change that in this change since it is completely orthogonal. It used to be useful to have them as 7.1 for dogfooding reasons, but not any more. We should fix this soon.

---
In reply to: [118551720](https://github.com/dotnet/roslyn/pull/19713#discussion_r118551720) [](ancestors = 118551720)

File: src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs (hunk 10 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs]: Good point. I ended up not using the refKind of "this", so I am not sure what code actually uses this. I will check if we actually need this change.

---
In reply to: [118552148](https://github.com/dotnet/roslyn/pull/19713#discussion_r118552148) [](ancestors = 118552148)

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs (hunk 8 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: We are invoking the "method" and methodContainingType is its owner type. We are not considering the caller here.  We are just trying to determine whether calee needs a writeable reference.  

---
In reply to: [118551237](https://github.com/dotnet/roslyn/pull/19713#discussion_r118551237) [](ancestors = 118551237)

File: src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs (hunk 10 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs]: The RefKind of "this" is used only in data flow analysis and that only cares about it being "Out" in a struct ctor and not "Out" in regular members.

Basically the change makes it to return a correct value (outside of the constructor), but none of the callers would seem to care about that.


---
In reply to: [118565720](https://github.com/dotnet/roslyn/pull/19713#discussion_r118565720) [](ancestors = 118565720,118552148)

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs]: To answer other questions - The `_containingType/_containingMethod` parts did not change. I just refactored the original code. 
Lambdas/LocalFunctions do not have ThisParameter, intentionally. Noone is supposed to get ThisParameter from a lambda symbol.

---
In reply to: [118576819](https://github.com/dotnet/roslyn/pull/19713#discussion_r118576819) [](ancestors = 118576819,118565720,118552148)

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Symbols/Source/ThisParameterSymbol.cs]: Still thinking if this can be tested in some reasonable scenario, other than just asking someMethod.ThisParameter.RefKind  directly. I guess, I'll add just that.

---
In reply to: [118578857](https://github.com/dotnet/roslyn/pull/19713#discussion_r118578857) [](ancestors = 118578857,118576819,118565720,118552148)

File: src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs [hunk 11]
```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs]: This file is mis-named "RadOnlyStructs.cs" instead of "ReadOnlyStructs.cs".

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 12]
```diff
@@ -174,6 +174,9 @@
   <data name="IDS_FeatureStaticClasses" xml:space="preserve">
     <value>static classes</value>
   </data>
+  <data name="IDS_FeatureReadonlyStructs" xml:space="preserve">
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Nit: if we're going with ReadOnly capitalization, it would apply to IDS_... too.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 13]
```diff
@@ -1601,8 +1601,18 @@ private MemberDeclarationSyntax ParseTypeDeclaration(SyntaxListBuilder<Attribute
 
                     return this.ParseClassOrStructOrInterfaceDeclaration(attributes, modifiers);
 
-                case SyntaxKind.RefKeyword:
                 case SyntaxKind.StructKeyword:
+                    // report use of ref struct class
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Is this the intended comment? I didn't understand what it means.

File: src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs [hunk 14]
```diff
@@ -249,6 +249,14 @@ internal sealed override bool IsByRefLikeType
             }
         }
 
+        internal sealed override bool IsReadOnly
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs]: We should probably add tests or update existing tests to verify this on various kinds of symbols.

**Comment by jcouv** [comment]
> // Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Same nit on name of file "ReadOnly". :-)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs:1 in cb1167c. [](commit_id = cb1167cec1b4d8a44c9ab3daa0c8c0b019adb2cf, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 13 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: should say readonly struct

---
In reply to: [118581738](https://github.com/dotnet/roslyn/pull/19713#discussion_r118581738) [](ancestors = 118581738)

**Comment by VSadov** [comment]
> // Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

I think this is fixed, but Git/codeflow does not show the diff 

---
In reply to: [304121824](https://github.com/dotnet/roslyn/pull/19713#issuecomment-304121824) [](ancestors = 304121824)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs:1 in cb1167c. [](commit_id = cb1167cec1b4d8a44c9ab3daa0c8c0b019adb2cf, deletion_comment = False)

**Comment by VSadov** [comment]
CI machines seem to be failing with IO errors.

**Comment by VSadov** [comment]
test this please

**Comment by VSadov** [comment]
test this please

**Comment by VSadov** [comment]
test this please

File: src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyStructPartialErr()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RadOnlyStructs.cs]: Should there be errors?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
+            );
+        }
+
+        [Fact]
+        public void ReadOnlyStructPartialErr()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: Should there be errors?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs (hunk 16 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: Should there be errors?

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 17]
```diff
@@ -398,7 +398,13 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
 
                     // SPEC: When this is used in a primary-expression within an instance method or instance accessor
                     // SPEC: of a struct, it is classified as a variable. 
-                    if (!thisref.Type.IsValueType || RequiresReturnableReference(valueKind))
+
+                    // Note: RValueOnly is checked at the beginning of this method. Since we are here we need more than readable.
+                    //"this" is only readable in members of readonly structs, unless we are in a constructor.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: >only readable [](start = 32, length = 13)

Consider replacing "only readable" with "read-only", I believe that would make the comment clearer.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs (hunk 16 above)
Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: Actually I did expect errors here, but we seem to not generally require modifiers to match. - You can state "unsafe" or "static" only on one part to have effect on all other parts. That could even be be intentional to support some source generation scenarios where generated parts would not need to restate modifiers. 

I will enter a bug to figure what to do with "ref" and "readonly", but will keep the behavior as is for now. And will change the name of the test as based on precedent I am no longer convinced if this should be an error.

---
In reply to: [118633262](https://github.com/dotnet/roslyn/pull/19713#discussion_r118633262) [](ancestors = 118633262)

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 18]
```diff
@@ -1412,26 +1412,38 @@ private void EmitCallExpression(BoundCall call, UseKind useKind)
...
-
-                        // calling a method defined in a value type
-                        Debug.Assert(receiverType == methodContainingType);
-                        // method is defined in the struct itself and is assumed to be mutating.
-                        //PROTOTYPE(readonlyRefs): when readonly structs are implemented, this will have to check "this"
-                        tempOpt = EmitReceiverRef(receiver, AddressKind.Writeable);
-                        callKind = CallKind.Call;
+                        // if method is defined in the struct itself it is assumed to be mutating, unless 
+                        // it is a member of a readonly struct and is not a constructor
+                        var thisAddresskind = methodContainingType.IsReadOnly && method.MethodKind != MethodKind.Constructor ?
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: >thisAddresskind [](start = 28, length = 15)

Why are we making an assumption that the receiver is ```this```. Perhaps the local should be renamed?

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 19]
```diff
@@ -2707,27 +2732,13 @@ private void EmitConstantExpression(TypeSymbol type, ConstantValue constantValue
         private void EmitInitObj(TypeSymbol type, bool used, SyntaxNode syntaxNode)
         {
             if (used)
-            {
-                if (type.IsPointerType() || type.SpecialType == SpecialType.System_UIntPtr)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: >if (type.IsPointerType() || type.SpecialType == SpecialType.System_UIntPtr) [](start = 16, length = 75)

It looks like many consumers of EmitInitObj will not be able to take advantage of this optimization. Is this intentional for all of them?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 20]
```diff
@@ -1601,8 +1601,18 @@ private MemberDeclarationSyntax ParseTypeDeclaration(SyntaxListBuilder<Attribute
 
                     return this.ParseClassOrStructOrInterfaceDeclaration(attributes, modifiers);
 
-                case SyntaxKind.RefKeyword:
                 case SyntaxKind.StructKeyword:
+                    // report use of "readonly struct" if feature is unsupported
+                    for (int i = 0, n = modifiers.Count; i < n; i++)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >for (int i = 0, n = modifiers.Count; i < n; i++) [](start = 20, length = 48)

It feels like this for can be extracted into a helper method that takes modifier kind and feature id, and shared between case SyntaxKind.ClassKeyword and case SyntaxKind.StructKeyword.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 21]
```diff
@@ -398,7 +398,13 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
...
+                    //"this" is not returnable by reference in a struct.
+                    if (!thisref.Type.IsValueType ||
+                        (thisref.Type.IsReadOnly && (this.ContainingMember() as MethodSymbol)?.MethodKind != MethodKind.Constructor) ||
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: >ContainingMember [](start = 58, length = 16)

Should this be ContainingMemberOrLambda?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs [hunk 22]
```diff
@@ -0,0 +1,522 @@
...
+        readonly int x;
+
+        public S1(int i)
+        {            
+            x = i; // OK
+        }
+
+        public S1(S1 arg)
+        {
+            this = arg; // OK
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs]: >this = arg; // OK [](start = 12, length = 17)

Consider adding a test that does this inside a lambda or a local function.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs]: > this = arg; // OK [](start = 11, length = 18)

I think we should have a test that runs the code to verify runtime behavior and probably IL as well.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs [hunk 23]
```diff
@@ -0,0 +1,522 @@
...
+    static void TakesRef(ref S1 arg){}
+    static void TakesRef(ref int arg){}
+
+    readonly struct S1
+    {
+        readonly int x;
+
+        public S1(int i)
+        {            
+            x = i; // OK
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenReadonlyStructTests.cs]: >x = i; // OK [](start = 12, length = 12)

I think we should have a test that runs the code to verify runtime behavior and probably IL as well.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+    readonly struct S1{}
+
+    public readonly struct S2{}
+
+    readonly public struct S3{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: >comp.VerifyDiagnostics( [](start = 12, length = 23)

Consider asserting relevant properties on symbols.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+    readonly class S1{}
+
+    public readonly delegate ref readonly int S2();
+
+    readonly public interface S3{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll);
+            comp.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: >comp.VerifyDiagnostics( [](start = 12, length = 23)

Consider asserting relevant properties on symbols.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+class Program
+{
+    readonly ref struct S1{}
+
+    unsafe readonly public ref struct S2{}
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(text, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.Latest), options: TestOptions.DebugDll.WithAllowUnsafe(true));
+            comp.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: >comp.VerifyDiagnostics( [](start = 12, length = 23)

Consider asserting relevant properties on symbols.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+        [WorkItem(19808, "https://github.com/dotnet/roslyn/issues/19808")]
+        [Fact]
+        public void ReadOnlyStructPartialNotMatchingModifiers()
+        {
+            var text = @"
+class Program
+{
+    readonly partial struct S1{}
+
+    readonly partial ref struct S1{}
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: >readonly [](start = 4, length = 8)

A scenario when this modifier is removed is probably interesting too.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 21 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Yes. it should be. I assumed that we can assign to initonly fields in methods nested in constructors. It is not actually the case (as it should be, for lambdas in any case). Will change and add tests. Good catch!

---
In reply to: [118784631](https://github.com/dotnet/roslyn/pull/19713#discussion_r118784631) [](ancestors = 118784631)

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs (hunk 27 above)
Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: how a modifier can be removed? 

---
In reply to: [118787804](https://github.com/dotnet/roslyn/pull/19713#discussion_r118787804) [](ancestors = 118787804)

Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReadOnlyStructs.cs]: ah, I see - mismatch in "readonly". Can add.

---
In reply to: [118796435](https://github.com/dotnet/roslyn/pull/19713#discussion_r118796435) [](ancestors = 118796435,118787804)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 20 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: will try.

---
In reply to: [118783089](https://github.com/dotnet/roslyn/pull/19713#discussion_r118783089) [](ancestors = 118783089)

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs (hunk 19 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: The purpose is that EmitInitObj does just that - emits "initobj". 
The optimizations will be done in the InitDefaultValue. We should do all the optimizations in one place.

If callers can benefit from the optimization they should call that. 

I will follow up on all the callers in a separate change. We were likely missing some optimizations already since some of them were already done only in InitDefaultValue. Did not want to mix unrelated change into this since it could affect unrelated testcases.

---
In reply to: [118782391](https://github.com/dotnet/roslyn/pull/19713#discussion_r118782391) [](ancestors = 118782391)

Comment (VSadov) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: Will add "PROTOTYPE" for this. There are several "PROTOTYPE" in codegen already so this one would be addressed together with others.

---
In reply to: [118798124](https://github.com/dotnet/roslyn/pull/19713#discussion_r118798124) [](ancestors = 118798124,118782391)

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs (hunk 18 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: receiverAddressKind seems a better name.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 21 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Well, turns out "this" is not accessible at all inside nested methods in structs. Not even for reading or any other kind of use.
I will change to ContainingMemberOrLambda, since it seems to be the right thing to do, and add related tests, but the difference would not observable right now.

---
In reply to: [118796219](https://github.com/dotnet/roslyn/pull/19713#discussion_r118796219) [](ancestors = 118796219,118784631)

**Comment by VSadov** [comment]
All feedback addressed.

**Comment by VSadov** [comment]
Test this please

**Comment by VSadov** [comment]
Simething wrong with jenkins again?

**Comment by VSadov** [comment]
Test ubuntu_14_debug_prtest please

**Comment by VSadov** [comment]
Test microbuild_prtest please 
Grrrrre

**Comment by VSadov** [comment]
Test microbuild_prtest please

**Comment by VSadov** [comment]
Test microbuild_prtest please

File: src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs [hunk 28]
```diff
@@ -2673,7 +2684,21 @@ private void EmitDefaultValue(TypeSymbol type, bool used, SyntaxNode syntaxNode)
                     }
                 }
 
-                EmitInitObj(type, true, syntaxNode);
+                if (type.IsPointerType() || type.SpecialType == SpecialType.System_UIntPtr)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: I didn't understand how this change relates to the PR. Was it covered in tests with VerifyIL?

Comment (VSadov) [inline src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs]: That is mostly refactoring of some optimizations to EmitDefaultValue from EmitInitObj.
We had existing tests for some cases and new tests that do something like `default<S1>.Method()` - that is how I noticed that we do some optimizations in one method and some in another - making method's responsibilities unclear.

**Comment by VSadov** [comment]
Thanks!!
